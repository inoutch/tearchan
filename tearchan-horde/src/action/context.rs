use crate::action::manager::TimeMilliseconds;

pub struct ActionContext {
    pub last_time: TimeMilliseconds, // The last time of stacking all actions
    pub running_end_time: TimeMilliseconds, // The end time of running action
    pub state_len: usize,            // Running and pending state length of entity
}
