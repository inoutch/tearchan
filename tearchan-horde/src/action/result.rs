use crate::action::Action;
use std::sync::Arc;

pub enum ActionResult<T> {
    Start {
        action: Arc<Action<T>>,
    },
    Update {
        action: Arc<Action<T>>,
        current_time: u64,
    },
    End {
        action: Arc<Action<T>>,
    },
}

impl<T> ActionResult<T> {
    pub fn get_start(&self) -> Option<&Arc<Action<T>>> {
        match self {
            ActionResult::Start { action } => Some(action),
            ActionResult::Update { .. } => None,
            ActionResult::End { .. } => None,
        }
    }

    pub fn get_update(&self) -> Option<(&Arc<Action<T>>, u64)> {
        match self {
            ActionResult::Start { .. } => None,
            ActionResult::Update {
                action,
                current_time,
            } => Some((action, *current_time)),
            ActionResult::End { .. } => None,
        }
    }

    pub fn get_end(&self) -> Option<&Arc<Action<T>>> {
        match self {
            ActionResult::Start { .. } => None,
            ActionResult::Update { .. } => None,
            ActionResult::End { action } => Some(action),
        }
    }
}
