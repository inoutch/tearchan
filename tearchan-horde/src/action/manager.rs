use crate::action::context::ActionContext;
use crate::action::result::ActionResult;
use crate::action::Action;
use crate::HordeInterface;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::rc::Rc;
use tearchan_ecs::component::EntityId;
use tearchan_util::btree::DuplicatableBTreeMap;

pub type TimeMilliseconds = u64;

pub struct ActionManager<T>
where
    T: HordeInterface,
{
    // Do serialize
    pending_actions: DuplicatableBTreeMap<TimeMilliseconds, Rc<Action<T::ActionState>>>,
    running_actions: DuplicatableBTreeMap<TimeMilliseconds, Rc<Action<T::ActionState>>>,
    current_time: TimeMilliseconds,
    // Cache and using for server only
    contexts: HashMap<EntityId, ActionContext>,
    pending_cache: HashSet<EntityId>,
}

impl<T> Default for ActionManager<T>
where
    T: HordeInterface,
{
    fn default() -> Self {
        ActionManager {
            pending_actions: DuplicatableBTreeMap::default(),
            running_actions: DuplicatableBTreeMap::default(),
            current_time: 0,
            contexts: HashMap::new(),
            pending_cache: HashSet::new(),
        }
    }
}

impl<T> ActionManager<T>
where
    T: HordeInterface,
{
    pub fn new(data: ActionManagerData<T::ActionState>) -> ActionManager<T> {
        let mut pending_actions = DuplicatableBTreeMap::default();
        let mut running_actions = DuplicatableBTreeMap::default();
        let mut contexts = HashMap::new();
        let mut pending_cache = HashSet::new();

        for entity_id in data.entity_ids {
            contexts.insert(
                entity_id,
                ActionContext {
                    last_time: data.current_time,
                    state_len: 0,
                },
            );
        }

        for action in data.actions {
            let context = contexts
                .entry(action.entity_id)
                .or_insert_with(|| ActionContext {
                    last_time: action.end_time,
                    state_len: 0,
                });
            context.state_len += 1;
            if context.last_time < action.end_time {
                context.last_time = action.end_time;
            }

            if data.current_time >= action.start_time {
                running_actions.push_back(action.start_time, action);
            } else {
                pending_actions.push_back(action.start_time, action);
            }
        }

        for (entity_id, context) in &contexts {
            if context.state_len == 0 {
                pending_cache.insert(*entity_id);
            }
        }

        ActionManager {
            pending_actions,
            running_actions,
            current_time: data.current_time,
            contexts,
            pending_cache,
        }
    }

    pub fn update(&mut self, elapsed_time: TimeMilliseconds) {
        self.current_time += elapsed_time;
    }

    pub fn is_running(&self, entity_id: EntityId) -> bool {
        match self.contexts.get(&entity_id) {
            Some(context) => context.state_len != 0,
            None => false,
        }
    }

    pub fn push_states(
        &mut self,
        entity_id: EntityId,
        states: Vec<(T::ActionState, TimeMilliseconds)>,
    ) {
        for (state, duration) in states {
            let last_time = self.get_context_mut(entity_id).last_time;
            let end_time = last_time + duration;
            self.pending_actions.push_back(
                last_time,
                Rc::new(Action::new(entity_id, last_time, end_time, state)),
            );
            self.get_context_mut(entity_id).last_time = end_time;
        }
    }

    pub fn pull(&mut self) -> DuplicatableBTreeMap<u64, ActionResult<T>> {
        let mut results = DuplicatableBTreeMap::default();
        while let Some(action) = self.pending_actions.pop_first_back() {
            if action.start_time() > self.current_time {
                self.pending_actions.push_front(action.start_time(), action);
                break;
            }

            let entity_id = action.entity_id();
            let start_time = action.start_time();
            let end_time = action.end_time();
            if action.end_time() > self.current_time {
                self.running_actions
                    .push_front(end_time, Rc::clone(&action));
                results.push_back(start_time, ActionResult::Start { action });

                let context = self.get_context_mut(entity_id);
                context.state_len += 1;
            } else {
                results.push_back(
                    start_time,
                    ActionResult::Start {
                        action: Rc::clone(&action),
                    },
                );
                results.push_back(end_time, ActionResult::End { action });

                let context = self.get_context_mut(entity_id);
                if context.state_len == 0 {
                    self.pending_cache.insert(entity_id);
                }
            }
        }

        while let Some(action) = self.running_actions.pop_first_back() {
            if action.end_time() > self.current_time {
                self.running_actions.push_front(action.end_time(), action);
                break;
            }

            let entity_id = action.entity_id();
            results.push_front(action.end_time(), ActionResult::End { action });

            let context = self.get_context_mut(entity_id);
            context.state_len -= 1;
            if context.state_len == 0 {
                self.pending_cache.insert(entity_id);
            }
        }

        for (_, started_stores) in self.running_actions.iter() {
            for action in started_stores.iter() {
                results.push_back(
                    self.current_time,
                    ActionResult::Update {
                        action: Rc::clone(action),
                        current_time: self.current_time,
                    },
                )
            }
        }
        results
    }

    /**
     * Get free entity ids after update_actions
     */
    pub fn clean_pending_entity_ids(&mut self) -> HashSet<EntityId> {
        std::mem::replace(&mut self.pending_cache, HashSet::new())
    }

    pub fn attach(&mut self, entity_id: EntityId) {
        debug_assert!(
            !self.contexts.contains_key(&entity_id),
            "entity of {} is already attached",
            entity_id
        );

        let context = ActionContext {
            last_time: self.current_time,
            state_len: 0,
        };
        self.contexts.insert(entity_id, context);
        self.pending_cache.insert(entity_id);
    }

    pub fn detach(&mut self, entity_id: EntityId) {
        self.contexts.remove(&entity_id);
        self.pending_cache.remove(&entity_id);
    }

    pub fn cancel(&mut self, entity_id: EntityId) {
        let context = self
            .contexts
            .get_mut(&entity_id)
            .expect("invalid entity_id");
        context.last_time = self.current_time;
        context.state_len = 0;
        self.pending_cache.insert(entity_id);

        for (_, pending_actions) in self.pending_actions.iter_mut() {
            pending_actions.retain(|action| action.entity_id != entity_id);
        }
        for (_, running_actions) in self.running_actions.iter_mut() {
            running_actions.retain(|action| action.entity_id != entity_id);
        }
    }

    pub fn create_data(&self) -> ActionManagerData<T::ActionState> {
        let mut actions = vec![];
        for (_, pending_actions) in self.pending_actions.iter() {
            for pending_action in pending_actions {
                actions.push(Rc::clone(&pending_action));
            }
        }
        for (_, running_actions) in self.running_actions.iter() {
            for running_action in running_actions {
                actions.push(Rc::clone(&running_action));
            }
        }

        ActionManagerData {
            actions,
            entity_ids: self
                .contexts
                .iter()
                .map(|(entity_id, _)| *entity_id)
                .collect(),
            current_time: self.current_time,
        }
    }

    fn get_context_mut(&mut self, entity_id: EntityId) -> &mut ActionContext {
        debug_assert!(
            self.contexts.contains_key(&entity_id),
            "entity of {} is not attached",
            entity_id
        );

        if self.contexts.get_mut(&entity_id).is_some() {
            return self.contexts.get_mut(&entity_id).unwrap();
        }

        let context = ActionContext {
            last_time: self.current_time,
            state_len: 0,
        };
        self.contexts.insert(entity_id, context);
        self.contexts.get_mut(&entity_id).unwrap()
    }
}

#[derive(Serialize, Deserialize)]
pub struct ActionManagerData<T> {
    pub actions: Vec<Rc<Action<T>>>,
    pub entity_ids: HashSet<EntityId>,
    pub current_time: TimeMilliseconds,
}
