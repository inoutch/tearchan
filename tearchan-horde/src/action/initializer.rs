use crate::action::manager::ActionManagerData;
use crate::ActionController;
use tearchan_ecs::component::EntityId;

pub enum ActionInitializer<'a, T> {
    Controller(ActionController<'a, T>),
    Data(&'a mut ActionManagerData<T>),
}

impl<'a, T> ActionInitializer<'a, T> {
    pub fn attach(&mut self, entity_id: EntityId) {
        match self {
            ActionInitializer::Controller(controller) => {
                controller.attach(entity_id);
            }
            ActionInitializer::Data(data) => {
                data.attach(entity_id);
            }
        }
    }

    pub fn detach(&mut self, entity_id: EntityId) {
        match self {
            ActionInitializer::Controller(controller) => {
                controller.detach(entity_id);
            }
            ActionInitializer::Data(data) => {
                data.detach(entity_id);
            }
        }
    }
}

impl<'a, T> From<ActionController<'a, T>> for ActionInitializer<'a, T> {
    fn from(controller: ActionController<'a, T>) -> Self {
        ActionInitializer::Controller(controller)
    }
}
