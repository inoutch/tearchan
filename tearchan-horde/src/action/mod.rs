use crate::action::manager::TimeMilliseconds;
use serde::{Deserialize, Serialize};
use std::rc::Rc;
use tearchan_ecs::component::EntityId;

pub mod context;
pub mod manager;
pub mod result;

#[derive(Debug, Serialize, Deserialize)]
pub struct Action<T> {
    entity_id: EntityId,
    start_time: TimeMilliseconds,
    end_time: TimeMilliseconds,
    inner: Rc<T>,
}

impl<T> Action<T> {
    pub fn new(
        entity_id: EntityId,
        start: TimeMilliseconds,
        end: TimeMilliseconds,
        inner: T,
    ) -> Action<T> {
        Action {
            entity_id,
            start_time: start,
            end_time: end,
            inner: Rc::new(inner),
        }
    }
}

impl<T> Action<T> {
    pub fn entity_id(&self) -> EntityId {
        self.entity_id
    }

    pub fn start_time(&self) -> TimeMilliseconds {
        self.start_time
    }

    pub fn end_time(&self) -> TimeMilliseconds {
        self.end_time
    }

    pub fn inner(&self) -> &Rc<T> {
        &self.inner
    }
}
