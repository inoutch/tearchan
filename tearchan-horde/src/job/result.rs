use crate::action::manager::TimeMilliseconds;

pub type ProgressState<ActionState> = (ActionState, TimeMilliseconds);

#[derive(Debug)]
pub struct JobResult<Job, ActionState> {
    pub creators: Vec<Job>,
    pub states: Vec<ProgressState<ActionState>>,
}

impl<ActionCreator, ActionState> Default for JobResult<ActionCreator, ActionState> {
    fn default() -> Self {
        JobResult {
            creators: vec![],
            states: vec![],
        }
    }
}
