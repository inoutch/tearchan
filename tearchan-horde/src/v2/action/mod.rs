use crate::action::manager::TimeMilliseconds;
use crate::v2::Tick;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tearchan_ecs::component::EntityId;

pub mod collection;
pub mod manager;

pub const VALID_SESSION_ID: ActionSessionId = 0;

pub type ActionSessionId = u64;

#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
pub enum ActionType {
    Start {
        tick: Tick,
    },
    Update {
        start: TimeMilliseconds,
        end: TimeMilliseconds,
    },
    End {
        tick: Tick,
    },
}

impl ActionType {
    pub fn tick(&self) -> Option<Tick> {
        match self {
            ActionType::Start { tick, .. } => Some(*tick),
            ActionType::Update { .. } => None,
            ActionType::End { tick, .. } => Some(*tick),
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Action<T> {
    raw: T,
    entity_id: EntityId,
    ty: ActionType,
}

impl<T> Action<T> {
    pub fn new(raw: T, entity_id: EntityId, ty: ActionType) -> Self {
        Self { raw, entity_id, ty }
    }

    pub fn tick(&self) -> Option<Tick> {
        self.ty.tick()
    }
}

pub type ArcAction<T> = Action<Arc<T>>;

impl<T> Action<T> {
    pub fn raw(&self) -> &T {
        &self.raw
    }

    pub fn replace<U>(&self, raw: U) -> Action<U> {
        Action {
            raw,
            entity_id: self.entity_id,
            ty: self.ty,
        }
    }

    pub fn entity_id(&self) -> EntityId {
        self.entity_id
    }
}
