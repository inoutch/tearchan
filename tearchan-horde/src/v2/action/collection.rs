use crate::v2::action::manager::ActionSessionValidator;
use crate::v2::action::{ActionSessionId, ArcAction};
use crate::v2::Tick;
use std::any::{Any, TypeId};
use std::collections::hash_map::Iter;
use std::collections::HashMap;
use tearchan_ecs::component::EntityId;

pub struct ActionMeta {
    pub type_id: TypeId,
    pub entity_id: EntityId,
    pub session_id: ActionSessionId,
    pub tick: Tick,
}

pub struct ActionAnyVec {
    metas: Vec<ActionMeta>,
    actions: Box<dyn Any>,
}

impl ActionAnyVec {
    pub fn new<T: 'static>() -> Self {
        ActionAnyVec {
            metas: vec![],
            actions: Box::new(Vec::<ArcAction<T>>::new()),
        }
    }
}

impl ActionAnyVec {
    pub fn cast<T: 'static>(&self, validator: &ActionSessionValidator) -> Vec<&ArcAction<T>> {
        self.actions
            .downcast_ref::<Vec<ArcAction<T>>>()
            .expect("Invalid type")
            .iter()
            .zip(self.metas.iter())
            .filter_map(|(action, meta)| {
                if validator.validate(meta) {
                    Some(action)
                } else {
                    None
                }
            })
            .collect()
    }

    pub fn cast_cloned<T: 'static>(&self, validator: &ActionSessionValidator) -> Vec<ArcAction<T>> {
        self.actions
            .downcast_ref::<Vec<ArcAction<T>>>()
            .expect("Invalid type")
            .iter()
            .zip(self.metas.iter())
            .filter_map(|(action, meta)| {
                if validator.validate(meta) {
                    Some(action.clone())
                } else {
                    None
                }
            })
            .collect()
    }

    fn push<T: 'static>(&mut self, action: ArcAction<T>, session_id: ActionSessionId, tick: Tick) {
        let type_id = TypeId::of::<T>();
        self.metas.push(ActionMeta {
            type_id,
            entity_id: action.entity_id,
            session_id,
            tick,
        });
        self.actions
            .downcast_mut::<Vec<ArcAction<T>>>()
            .expect("Invalid type")
            .push(action);
    }
}

#[derive(Default)]
pub struct TypedActionAnyMap {
    map: HashMap<TypeId, ActionAnyVec>,
}

impl TypedActionAnyMap {
    pub fn push<T>(&mut self, action: ArcAction<T>, session_id: ActionSessionId)
    where
        T: 'static,
    {
        let type_id = TypeId::of::<T>();
        let tick = action.tick().expect("Must have tick");
        self.map
            .entry(type_id)
            .or_insert_with(ActionAnyVec::new::<T>)
            .push(action, session_id, tick);
    }

    pub fn get<T>(&self, validator: &ActionSessionValidator) -> Option<Vec<&ArcAction<T>>>
    where
        T: 'static,
    {
        let vec = self.map.get(&TypeId::of::<T>())?.cast(validator);
        if vec.is_empty() {
            return None;
        }
        Some(vec)
    }

    pub fn get_cloned<T>(&self, validator: &ActionSessionValidator) -> Option<Vec<ArcAction<T>>>
    where
        T: 'static,
    {
        let vec = self.map.get(&TypeId::of::<T>())?.cast_cloned(validator);
        if vec.is_empty() {
            return None;
        }
        Some(vec)
    }

    pub fn iter(&self) -> Iter<'_, TypeId, ActionAnyVec> {
        self.map.iter()
    }
}

#[derive(Default)]
pub struct AnyActionVec {
    vec: Vec<Box<dyn Any>>,
}

impl AnyActionVec {
    pub fn cast<T: 'static>(&self) -> Vec<&ArcAction<T>> {
        self.vec
            .iter()
            .filter_map(|item| item.downcast_ref())
            .collect()
    }

    pub fn cast_cloned<T: 'static>(&self) -> Vec<ArcAction<T>> {
        self.vec
            .iter()
            .filter_map(|item| item.downcast_ref::<ArcAction<T>>())
            .cloned()
            .collect::<Vec<_>>()
    }

    pub fn push<T: 'static>(&mut self, item: T) {
        self.vec.push(Box::new(item));
    }

    #[inline]
    pub fn is_empty(&self) -> bool {
        self.vec.is_empty()
    }

    #[inline]
    pub fn len(&self) -> usize {
        self.vec.len()
    }

    #[inline]
    fn replace(&mut self, index: usize, item: Box<dyn Any>) {
        if let Some(x) = self.vec.get_mut(index) {
            *x = item
        }
    }

    #[inline]
    fn pop(&mut self) -> Option<Box<dyn Any>> {
        self.vec.pop()
    }

    #[inline]
    fn push_as_raw(&mut self, raw: Box<dyn Any>) {
        self.vec.push(raw);
    }
}

#[derive(Default)]
pub struct TypedAnyActionMapGroupedByEntityId {
    map: HashMap<TypeId, AnyActionVec>,
    indices: HashMap<EntityId, (TypeId, usize)>,
    entities: HashMap<(TypeId, usize), EntityId>,
}

impl TypedAnyActionMapGroupedByEntityId {
    pub fn get<T>(&self) -> Option<Vec<&ArcAction<T>>>
    where
        T: 'static,
    {
        Some(self.map.get(&TypeId::of::<T>())?.cast())
    }

    pub fn get_cloned<T>(&self) -> Option<Vec<ArcAction<T>>>
    where
        T: 'static,
    {
        Some(self.map.get(&TypeId::of::<T>())?.cast_cloned())
    }

    pub fn insert<T>(&mut self, entity_id: EntityId, action: ArcAction<T>)
    where
        T: 'static,
    {
        let type_id = TypeId::of::<T>();

        self.remove(entity_id);
        let vec = self
            .map
            .entry(type_id)
            .or_insert_with(AnyActionVec::default);
        let index = vec.len();
        vec.push(action);
        self.entities.insert((type_id, index), entity_id);
        self.indices.insert(entity_id, (type_id, index));
    }

    pub fn insert_with_type_id(
        &mut self,
        entity_id: EntityId,
        action: Box<dyn Any>,
        type_id: TypeId,
    ) {
        self.remove(entity_id);

        let collection = self
            .map
            .entry(type_id)
            .or_insert_with(AnyActionVec::default);
        let index = collection.len();
        collection.push_as_raw(action);
        self.entities.insert((type_id, index), entity_id);
        self.indices.insert(entity_id, (type_id, index));
    }

    pub fn remove(&mut self, entity_id: EntityId) {
        if let Some((type_id, index)) = self.indices.remove(&entity_id) {
            self.entities.remove(&(type_id, index));

            let collection = self.map.get_mut(&type_id).unwrap();
            let last = collection.len() - 1;
            let last_item = collection.pop().unwrap();
            if index != last {
                let last_entity_id = self.entities.remove(&(type_id, last)).unwrap();
                self.indices.remove(&last_entity_id);

                collection.replace(index, last_item);
                self.entities.insert((type_id, index), last_entity_id);
                self.indices.insert(last_entity_id, (type_id, index));
            } else if index == 0 {
                self.map.remove(&type_id);
            }
        }
    }

    pub fn iter(&self) -> Iter<'_, TypeId, AnyActionVec> {
        self.map.iter()
    }
}

#[cfg(test)]
mod test {
    use crate::v2::action::collection::TypedAnyActionMapGroupedByEntityId;
    use crate::v2::action::{Action, ActionType};
    use std::sync::Arc;

    #[derive(Clone, Debug)]
    struct MoveState;

    #[derive(Clone, Debug)]
    struct JumpState;

    #[test]
    fn test_typed_any_vec_grouped_by_entities() {
        let mut collection = TypedAnyActionMapGroupedByEntityId::default();

        collection.insert(
            1,
            Action {
                raw: Arc::new(MoveState),
                entity_id: 1,
                ty: ActionType::Start {
                    tick: 0,
                    start: 0,
                    end: 0,
                },
            },
        );

        collection.insert(
            2,
            Action {
                raw: Arc::new(JumpState),
                entity_id: 2,
                ty: ActionType::Start {
                    tick: 0,
                    start: 0,
                    end: 0,
                },
            },
        );
        collection.insert(
            2,
            Action {
                raw: Arc::new(JumpState),
                entity_id: 2,
                ty: ActionType::Start {
                    tick: 0,
                    start: 0,
                    end: 0,
                },
            },
        );

        insta::assert_debug_snapshot!(collection.get::<MoveState>());
        insta::assert_debug_snapshot!(collection.get::<JumpState>());

        collection.remove(2);

        insta::assert_debug_snapshot!(collection.get::<MoveState>()); // should have one
        insta::assert_debug_snapshot!(collection.get::<JumpState>()); // should be empty

        collection.remove(1);

        insta::assert_debug_snapshot!(collection.get::<MoveState>()); // should be empty

        assert_eq!(collection.entities.len(), 0);
        assert_eq!(collection.indices.len(), 0);
        assert_eq!(collection.map.len(), 0);
    }
}
