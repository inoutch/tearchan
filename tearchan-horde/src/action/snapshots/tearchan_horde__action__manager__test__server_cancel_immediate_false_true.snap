---
source: tearchan-horde/src/action/manager.rs
expression: convert_events(&mut action_manager)

---
[
    Enqueue {
        action: Action {
            entity_id: 1,
            start_time: 0,
            end_time: 0,
            inner: "Sleep",
        },
    },
    Enqueue {
        action: Action {
            entity_id: 1,
            start_time: 0,
            end_time: 3000,
            inner: "Walk",
        },
    },
    Start {
        action: Action {
            entity_id: 1,
            start_time: 0,
            end_time: 3000,
            inner: "Walk",
        },
    },
    Cancel {
        action: Action {
            entity_id: 1,
            start_time: 0,
            end_time: 3000,
            inner: "Walk",
        },
    },
    Start {
        action: Action {
            entity_id: 1,
            start_time: 0,
            end_time: 0,
            inner: "Sleep",
        },
    },
    Cancel {
        action: Action {
            entity_id: 1,
            start_time: 0,
            end_time: 0,
            inner: "Sleep",
        },
    },
]
